src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzer.cs

```diff
@@ -0,0 +1,171 @@
+                                                                             s_localizableOfTypeTitle,
+                                                                             s_localizableOfTypeMessage,
+                                                                             DiagnosticCategory.Reliability,
+                                                                             RuleLevel.BuildWarningCandidate,
+                                                                             description: s_localizableOfTypeDescription,
+                                                                             isPortedFxCopRule: false,
+                                                                             isDataflowRule: false);
+
+        private static readonly ImmutableDictionary<string, DiagnosticDescriptor> methodMetadataNames = new Dictionary<string, DiagnosticDescriptor>
+        {
+            ["OfType"] = OfTypeRule,
+            ["Cast"] = CastRule,
+        }.ToImmutableDictionary();
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+            => methodMetadataNames.Values.ToImmutableArray();
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(compilationStartContext =>
+            {
+                var enumerableType = compilationStartContext.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemLinqEnumerable);
+                var genericIEnumerableType = compilationStartContext.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericIEnumerable1);
+
+                if (enumerableType == null || genericIEnumerableType == null)
+                {
+                    return;
+                }
+
+                compilationStartContext.RegisterOperationAction(operationContext =>
+                {
+                    var invocation = (IInvocationOperation)operationContext.Operation;
+
+                    var targetMethod = invocation.TargetMethod.ReducedFrom ?? invocation.TargetMethod;
+                    if (!methodMetadataNames.TryGetValue(targetMethod.Name, out var rule))
+                    {
+                        return;
+                    }
+
+                    if (!invocation.TargetMethod.TypeParameters.HasExactly(1))
+                    {
+                        return;
+                    }
+
+                    var numberOfArguments = invocation.IsExtensionMethodAndHasNoInstance() ? 1 : 0;
+                    if (!invocation.Arguments.HasExactly(numberOfArguments))
+                    {
+                        return;
+                    }
+
+                    var arg = invocation.IsExtensionMethodAndHasNoInstance()
```

Comment:
I need to look for it but I am sure that we have something similar and that there is a little trick for vbnet.

Reply:
this does seem pretty messy. I remember looking for quite a while but couldn't find anything.

Reply:
found it! `GetInstance()`